Deploy-CAM.ps1

```diff
@@ -135,6 +180,48 @@ function Login-AzureRmAccountWithBetterReporting($Credential) {
     }
 }
 
+# uses session instance profile and TokenCache and returns an access token without having to authentication a second time
+function Get-AzureRmCachedAccessToken() {
+    $ErrorActionPreference = 'Stop'
+    if(-not (Get-Module AzureRm.Profile)) {
+        Import-Module AzureRm.Profile
+    }
+    $azureRmProfileModuleVersion = (Get-Module AzureRm.Profile).Version
+    
+    # refactoring performed in AzureRm.Profile v3.0 or later
+    if($azureRmProfileModuleVersion.Major -ge 3) {
+        $azureRmProfile = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile
+        if(-not $azureRmProfile.Accounts.Count) {
+            Write-Error "Ensure you have logged in before calling this function."
+        }
+    } else {
+        # AzureRm.Profile < v3.0
+        $azureRmProfile = [Microsoft.WindowsAzure.Commands.Common.AzureRmProfileProvider]::Instance.Profile
+        if(-not $azureRmProfile.Context.Account.Count) {
+            Write-Error "Ensure you have logged in before calling this function."
+        }
+    }
+    $currentAzureContext = Get-AzureRmContext
+    $profileClient = New-Object Microsoft.Azure.Commands.ResourceManager.Common.RMProfileClient($azureRmProfile)
+    Write-Debug ("Getting access token for tenant" + $currentAzureContext.Subscription.TenantId)
+    $token = $profileClient.AcquireAccessToken($currentAzureContext.Subscription.TenantId)
+    $token.AccessToken
+}
+
+function Get-OwnerUpn() {
+    try {
+		$accessToken = Get-AzureRmCachedAccessToken
+		$decodedToken = Get-DecodedJWT `
+			-Token $accessToken `
+
+		$decodedToken.claims.upn
+	}
+	catch {
+		$errorMessage = "An error occured while retrieving owner upn."
```

Comment:
Have you tried this in Cloud Shell as well? I don't know if the token in Cloud Shell contains a UPN or a GUID?

Reply:
I tried this in PowerShell and not in Cloud Shell. However, the place to retrieve the access token is important because the context changes after calling Deploy-CAM and the token after that doesn't contain UPN anymore. That's why I am calling Get-OwnerUpn right before Deploy-CAM gets called.

Reply:
Okay - I think we need to ensure success in Cloud Shell before merging.